R/test-module.R

```diff
@@ -49,16 +49,17 @@ testModule <- function(module, expr, args, ...) {
 .testModule <- function(module, expr, args, ...) {
   # Capture the environment from the module
   # Inserts `session$env <- environment()` at the top of the function
-  fn_body <- body(module)
-  fn_body[seq(3, length(fn_body)+1)] <- fn_body[seq(2, length(fn_body))]
-  fn_body[[2]] <- quote(session$env <- environment())
-  body(module) <- fn_body
+  body(module) <- rlang::expr({
+    session$env <- environment()
+    !!!body(module)
+  })
 
   # Create a mock session
   session <- MockShinySession$new()
 
   # Parse the additional arguments
-  args <- list(..., input = session$input, output = session$output, session = session)
+  args <- if (missing(args)) list() else args
```

Comment:
Using `missing()` is generally not a good idea because it makes it hard to programmatically call a function. I think it would be better to modify the signature of `testModule` so that it has `args = NULL`.

Reply:
A couple more thoughts:

* `args` should be renamed to `args_` to avoid name collisions (if the module itself takes `args`).
* We should validate that `args_` is either NULL, or a list with all elements named. The reason to check for all named elements is because if any are not named, the order that they'll be passed to the module function will be opaque to the user.

Reply:
New thoughts:

* Get rid of `args`, since the user can use `!!!` to splice in lists of args.
* Provide an example of using `!!!`.
* In the documentation, it says that `...` are "additional named arguments", but that should be changed to say "additional arguments".
* See what happens if user provides input/output/session vars.